src/ranking_internals.cpp

```diff
@@ -167,8 +167,17 @@ MEMKIND_EXPORT void Ranking::AddPages(uintptr_t start_addr, size_t nof_pages,
         auto page_it = page_it_inserted.first;
         bool inserted = page_it_inserted.second;
         assert(inserted && "page not constructed!");
+        
+        // add page to the hotnessToPages map
         PageMetadata *page = &page_it->second;
-        this->hotnessToPages.insert(std::make_pair(page->GetHotness(), page));
+        std::set<PageMetadata *> same_hotness_pages;
```

Comment:
I'm not sure how much overhead constructing a set gives... Have you compared total CPU usage for longer tests, not just redis IOPS ?

Oftentimes, we add a lot of pages of the same hotness, e.g. after mmap. Maybe we could add a function that only works for pages of the same hotness and constructs only a single set and use it whenever possible?

Reply:
I think that creating a set (with just a root node) which could be abandoned is much less of a burden than searching through the hotness map. In case when there is no element in the map with the given hotness, you would traverse the map twice - one time in a search for a key and the second time when inserting a new element to the map.